src/rvsdg/egglog_optimizer/subst.rs

```diff
@@ -42,12 +44,144 @@ fn subst_all_rule(btype: String) -> String {
+               (= above     (Theta from inputs outputs)))
+              ((union above (Theta to   inputs outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= above     (Theta pred inputs from)))
+              ((union above (Theta pred inputs to)))
+              :ruleset subst)
+        (rule ((can-subst-Operand-beneath above pred-from pred-to)
+               (can-subst-VecOperand-beneath above outputs-from outputs-to)
+               (= above     (Theta pred-from inputs outputs-from)))
+              ((union above (Theta pred-from inputs outputs-to)))
+              :ruleset subst)
+        (rule ((can-subst-VecVecOperand-beneath above from to)
+               (= above     (Gamma pred inputs from)))
+              ((union above (Gamma pred inputs to)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= above     (OperandGroup from)))
+              ((union above (OperandGroup to)))
+              :ruleset subst)
+
+        ;; Learn can-subst-Operand-beneath
+        (rule ((can-subst-Body-beneath above from to)
+               (= new-from (Node from)))
+              ((can-subst-Operand-beneath above new-from (Node to)))
+              :ruleset subst)
+        (rule ((can-subst-Body-beneath above from to)
+               (= new-from (Project i from)))
+              ((can-subst-Operand-beneath above new-from (Project i to)))
+              :ruleset subst)
+
+        ;; Learn can-subst-body-beneath
+        (rule ((can-subst-Expr-beneath above from to)
+               (= new-from (PureOp from)))
+              ((can-subst-Body-beneath above new-from (PureOp to)))
+              :ruleset subst)
+        ;; Propagates up same context (Gamma: pred & inputs, Theta: inputs)
+        ;; rtjoa: Is it sound to propagate up outputs if we renumber args?
+        (rule ((can-subst-Operand-beneath above from to)
+               (= new-from (Gamma from inputs outputs)))
+              ((can-subst-Body-beneath above new-from (Gamma to inputs outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= new-from (Gamma pred from outputs)))
+              ((can-subst-Body-beneath above new-from (Gamma pred to outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= new-from (Theta pred from outputs)))
+              ((can-subst-Body-beneath above new-from (Theta pred to outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= new-from (OperandGroup from)))
+              ((can-subst-Body-beneath above new-from (OperandGroup to)))
+              :ruleset subst)
+        "
+    .into()];
+
+    // Learn can-subst-expr-beneath
+    res.push(
+        "
+      (rule ((can-subst-VecOperand-beneath above from to)
```

Comment:
Why is this in a separate string? Can't it be in the big string above?

Reply:
I just wanted the comment in rust for consistency with the surrounding commends separating each `can-subst-TYPE-beneath`